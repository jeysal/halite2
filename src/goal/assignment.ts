import GameMap from 'halite/GameMap';
import { OwnShip } from 'halite/Ship';
import { maxBy } from 'ramda';

import Goal from '../goal/goal';

interface ShipGoal<Turn> {
  ship: OwnShip<Turn>;
  goal: Goal<Turn>;
}

const VALUE_IMPACT = 0.5;
const SUITABILITY_IMPACT = 0.5;

const shipGoalValue = <Turn>(
  ship: OwnShip<Turn>,
  goal: Goal<Turn>,
  gm: GameMap<Turn>,
): number =>
  VALUE_IMPACT * goal.value(gm) +
  SUITABILITY_IMPACT * goal.suitability(ship, gm);

const assignGoalsToShips = <Turn>(
  goals: Goal<Turn>[],
  gm: GameMap<Turn>,
): ShipGoal<Turn>[] => {
  const ownShips = gm.myShips;

  const shipGoalValues = ownShips.map(ship =>
    goals.map(goal => ({ ship, goal, val: shipGoalValue(ship, goal, gm) })),
  );

  const goalsWithShips = shipGoalValues.reduce((gwss, gVals) => {
    const gValsBelowLimit = gVals.filter(gVal => {
      const entry = gwss.find(gws => gws.goal === gVal.goal);
      return !entry || entry.ships.length < entry.goal.maxShips();
    });
    if (gValsBelowLimit.length) {
      const bestVal = gValsBelowLimit.reduce(maxBy(value => value.val));
      return gwss.map(
        goalWithShips =>
          bestVal && goalWithShips.goal === bestVal.goal
            ? {
                goal: goalWithShips.goal,
                ships: [...goalWithShips.ships, bestVal.ship],
              }
            : goalWithShips,
      );
    }
    return gwss;
  }, goals.map(goal => ({ goal, ships: [] as OwnShip<Turn>[] })));

  // TODO goal minima

  return goalsWithShips.reduce(
    (acc, gws) => [
      ...acc,
      ...gws.ships.map(ship => ({ ship, goal: gws.goal })),
    ],
    [] as ShipGoal<Turn>[],
  );
};

export default assignGoalsToShips;
