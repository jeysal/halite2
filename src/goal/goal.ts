import { Action } from 'halite/Game';
import GameMap from 'halite/GameMap';
import { OwnShip } from 'halite/Ship';

export default interface Goal<Turn = number> {
  value(gm: GameMap<Turn>): number;
  minShips(): number;
  maxShips(): number;
  suitability(ship: OwnShip<Turn>, gm: GameMap<Turn>): number;
  strive(ship: OwnShip<Turn>, gm: GameMap<Turn>): string & Action<Turn>;
};
