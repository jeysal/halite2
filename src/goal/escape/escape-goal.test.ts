import DockingStatus from 'halite/DockingStatus';
import GameMap from 'halite/GameMap';
import Ship, { OwnShip } from 'halite/Ship';

import { goalStats } from '../../test/goal';
import { emptyMap } from '../../test/util';

import EscapeGoal, { determineEscapeGoals } from './escape-goal';

describe('value', () => {
  test('is high if one enemy dominates the match and we are behind', () => {
    const gm = new GameMap({ myPlayerId: 0, width: 100, height: 100 });
    gm.addPlayerShips(0, [{}]);
    gm.addPlayerShips(1, [{}]);
    gm.addPlayerShips(2, Array(20).fill({}));

    expect(new EscapeGoal().value(gm)).toMatchSnapshot();
  });

  test('is high if we are the only losing party', () => {
    const gm = new GameMap({ myPlayerId: 0, width: 100, height: 100 });
    gm.addPlayerShips(0, [{}]);
    gm.addPlayerShips(1, Array(20).fill({}));
    gm.addPlayerShips(2, Array(20).fill({}));

    expect(new EscapeGoal().value(gm)).toMatchSnapshot();
  });

  test('is rather low if one enemy dominates the match, but we still stand a chance', () => {
    const gm = new GameMap({ myPlayerId: 0, width: 100, height: 100 });
    gm.addPlayerShips(0, Array(15).fill({}));
    gm.addPlayerShips(1, [{}]);
    gm.addPlayerShips(2, Array(25).fill({}));

    expect(new EscapeGoal().value(gm)).toMatchSnapshot();
  });

  test('is low if the match is even', () => {
    const gm = new GameMap({ myPlayerId: 0, width: 100, height: 100 });
    gm.addPlayerShips(0, Array(20).fill({}));
    gm.addPlayerShips(1, Array(20).fill({}));
    gm.addPlayerShips(2, Array(20).fill({}));

    expect(new EscapeGoal().value(gm)).toMatchSnapshot();
  });

  test('is very low if we dominate the match', () => {
    const gm = new GameMap({ myPlayerId: 0, width: 100, height: 100 });
    gm.addPlayerShips(0, Array(40).fill({}));
    gm.addPlayerShips(1, Array(20).fill({}));
    gm.addPlayerShips(2, [{}]);

    expect(new EscapeGoal().value(gm)).toMatchSnapshot();
  });

  test('is 0 if there is only two players remaining', () => {
    const gm = new GameMap({ myPlayerId: 0, width: 100, height: 100 });
    gm.addPlayerShips(0, [{}]);
    gm.addPlayerShips(1, Array(20).fill({}));

    expect(new EscapeGoal().value(gm)).toBe(0);
  });

  test('is 0 if there are few overall ships', () => {
    const gm = new GameMap({ myPlayerId: 0, width: 100, height: 100 });
    gm.addPlayerShips(0, [{}, {}, {}]);
    gm.addPlayerShips(1, [{}, {}, {}]);
    gm.addPlayerShips(2, [{}, {}, {}]);

    expect(new EscapeGoal().value(gm)).toBe(0);
  });
});

describe('minShips', () => {
  test('is 1', () => {
    expect(new EscapeGoal().minShips()).toBe(1);
  });
});
describe('maxShips', () => {
  test('is 3', () => {
    expect(new EscapeGoal().maxShips()).toBe(3);
  });
});

describe('suitability', () => {
  test('is high for a remote ship', () => {
    expect(
      new EscapeGoal().suitability(
        new Ship(emptyMap, 0, { x: 15, y: 15 }) as OwnShip,
        emptyMap,
      ),
    ).toMatchSnapshot();
  });

  test('is rather low for a central ship', () => {
    expect(
      new EscapeGoal().suitability(
        new Ship(emptyMap, 0, { x: 35, y: 35 }) as OwnShip,
        emptyMap,
      ),
    ).toMatchSnapshot();
  });
});

describe('strive', () => {
  test('navigates to the closest corner of the map', () => {
    expect(
      new EscapeGoal().strive(
        new Ship(emptyMap, 0, { id: 100, x: 90, y: 90 }) as OwnShip,
        emptyMap,
      ),
    ).toMatchSnapshot();
  });

  test('undocks a docked ship', () => {
    expect(
      new EscapeGoal().strive(
        new Ship(emptyMap, 0, {
          id: 100,
          x: 99,
          y: 99,
          dockingStatus: DockingStatus.DOCKED,
        }) as OwnShip,
        emptyMap,
      ),
    ).toMatchSnapshot();
  });
});

describe('determineEscapeGoals', () => {
  test('creates a single escape goal', () => {
    expect(
      determineEscapeGoals().map(goal =>
        goalStats(goal, new Ship(emptyMap, 0, { x: 15, y: 15 }), emptyMap),
      ),
    );
  });
});
