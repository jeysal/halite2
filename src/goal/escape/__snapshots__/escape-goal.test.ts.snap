// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`strive navigates to the closest corner of the map 1`] = `"t 100 6 45"`;

exports[`strive undocks a docked ship 1`] = `"u 100"`;

exports[`suitability is high for a remote ship 1`] = `0.9146912192286945`;

exports[`suitability is rather low for a central ship 1`] = `0.7400828044922853`;

exports[`value is high if one enemy dominates the match and we are behind 1`] = `1`;

exports[`value is high if we are the only losing party 1`] = `1`;

exports[`value is low if the match is even 1`] = `0.2857142857142857`;

exports[`value is rather low if one enemy dominates the match, but we still stand a chance 1`] = `0.5714285714285714`;

exports[`value is very low if we dominate the match 1`] = `0.03571428571428571`;
